### PR TITLE
Add support for dual stack clusters

### PR DESCRIPTION
**Pull request by UXabre**
Adds support for dual stack clusters

resolves: https://github.com/hashicorp/vault-helm/issues/832

### CODE REVIEW

**Comment by M0NsTeRRR** [comment]
Looking for this feature too, as I see on other PR @UXabre maybe you can ask for a review of hashicorp team to push it further ?

**Comment by UXabre** [comment]
I've updated this PR to resolve the conflict, if somebody could review this? :-)

**Comment by hsimon-hashicorp** [comment]
> @hsimon-hashicorp Maybe you can help me with this PR? As I see nobody has reviewed it yet and perhaps you are my last hope :-) (sorry for spamming)

Not spam at all, this is my job. :) I've pinged our engineers to see if one can help with a review. Thanks!

**Comment by belyaev-dev** [comment]
joining waitlist :)

**Review by tvoran — COMMENTED** [review]
Thanks, left a couple thoughts. And should we set these parameters in `templates/server-headless-service.yaml` as well?

File: test/unit/server-service.bats [hunk 1]
```diff
@@ -467,3 +467,23 @@ load _helpers
...
+
+@test "server/Service: Assert ipFamilyPolicy set" {
+  cd `chart_dir`
+  local actual=$(helm template \
+      --show-only templates/server-service.yaml \
+      --set 'server.service.ipFamilyPolicy=PreferDualStack' \
+      . | tee /dev/stderr |
+      yq -r '.spec.ipFamilyPolicy' | tee /dev/stderr)
+  [ "${actual}" = "PreferDualStack" ]
+}
```
[the remark this answers is not in the document]
Comment (UXabre) [reply test/unit/server-service.bats:511]: Added these tests as requested

File: values.yaml [hunk 2]
```diff
@@ -676,6 +676,23 @@ server:
...
+    #     SingleStack: Single-stack service. The control plane allocates a cluster IP for the Service, using the first configured service cluster IP range.
+    #     PreferDualStack: Allocates IPv4 and IPv6 cluster IPs for the Service.
+    #     RequireDualStack: Allocates Service .spec.ClusterIPs from both IPv4 and IPv6 address ranges.
+    # ipFamilyPolicy: SingleStack
+    # 
+    # Sets the families that should be supported and the order in which they should be applied to ClusterIP as well
+    #
+    # ipFamilies:
+    # - IPv4
+
```
[the remark this answers is not in the document]
Comment (UXabre) [reply values.yaml:694]: Changed the defaults to these proposed

**Comment by UXabre** [comment]
> Thanks, left a couple thoughts. And should we set these parameters in `templates/server-headless-service.yaml` as well?

I've also added this to headless :-)

**Comment by UXabre** [comment]
All changes have been applied and I've rebased with main just to be on the safe side :-)

File: templates/ui-service.yaml [hunk 3]
```diff
@@ -20,6 +20,15 @@ metadata:
     app.kubernetes.io/managed-by: {{ .Release.Service }}
   {{- template "vault.ui.annotations" . }}
 spec:
+
```
Comment (tvoran) [inline templates/ui-service.yaml:23]:
```suggestion
```

File: values.yaml [hunk 4]
```diff
@@ -676,6 +676,22 @@ server:
     # or NodePort.
     #type: ClusterIP
 
+    # The IP family and IP families options are to set the behaviour in a dual-stack environment
+    # Omitting these values will let the service fall back to whatever the CNI dictates the defaults
+    # should be
```
Comment (tvoran) [inline values.yaml:681]:
```suggestion
    # The IP family and IP families options are to set the behaviour in a dual-stack environment.
    # Omitting these values will let the service fall back to whatever the CNI dictates the defaults
    # should be.
```

File: values.yaml [hunk 5]
```diff
@@ -676,6 +676,22 @@ server:
...
+    #
+    # Configures the service's supported IP family policy, can be either:
+    #     SingleStack: Single-stack service. The control plane allocates a cluster IP for the Service, using the first configured service cluster IP range.
+    #     PreferDualStack: Allocates IPv4 and IPv6 cluster IPs for the Service.
+    #     RequireDualStack: Allocates Service .spec.ClusterIPs from both IPv4 and IPv6 address ranges.
+    ipFamilyPolicy: ""
+    # 
+    # Sets the families that should be supported and the order in which they should be applied to ClusterIP as well
+    #
+    ipFamilies: []
```
Comment (tvoran) [inline values.yaml:693]:
```suggestion
    # Sets the families that should be supported and the order in which they should be applied to ClusterIP as well.
    # Can be IPv4 and/or IPv6
    ipFamilies: []
```

File: values.yaml [hunk 6]
```diff
@@ -979,6 +995,21 @@ ui:
...
+  # Sets the families that should be supported and the order in which they should be applied to ClusterIP as well
+  # Can be IPv4 and/or IPv6
+  serviceIpFamilies: []
+
+  # Configures the service's supported IP family, can be either:
+  #     SingleStack: Single-stack service. The control plane allocates a cluster IP for the Service, using the first configured service cluster IP range.
+  #     PreferDualStack: Allocates IPv4 and IPv6 cluster IPs for the Service.
+  #     RequireDualStack: Allocates Service .spec.ClusterIPs from both IPv4 and IPv6 address ranges.
+  # Only supported for kubernetes versions >=1.23.0
+  serviceIpFamilyPolicy: ""
```
Comment (tvoran) [inline values.yaml:1011]: Minor things, but I'd suggest making the order of the parameters and the doc strings match what's in `server.service`, and also uppercase the "IP" in "serviceIP...".

```suggestion
  # The IP family and IP families options are to set the behaviour in a dual-stack environment.
  # Omitting these values will let the service fall back to whatever the CNI dictates the defaults
  # should be.
  #
  # Only supported for kubernetes versions >=1.23.0
  #
  # Configures the service's supported IP family, can be either:
  #     SingleStack: Single-stack service. The control plane allocates a cluster IP for the Service, using the first configured service cluster IP range.
  #     PreferDualStack: Allocates IPv4 and IPv6 cluster IPs for the Service.
  #     RequireDualStack: Allocates Service .spec.ClusterIPs from both IPv4 and IPv6 address ranges.
  serviceIPFamilyPolicy: ""
  #
  # Sets the families that should be supported and the order in which they should be applied to ClusterIP as well.
  # Can be IPv4 and/or IPv6
  serviceIPFamilies: []
```

File: templates/ui-service.yaml (hunk 3 above)
Comment (UXabre) [reply templates/ui-service.yaml]: Removed empty line

File: values.yaml (hunk 4 above)
Comment (UXabre) [reply values.yaml]: Added punctuation

File: values.yaml (hunk 6 above)
Comment (UXabre) [reply values.yaml]: Yes, I agree, it should be consistent! I've made the changes as requested

File: values.yaml (hunk 5 above)
Comment (UXabre) [reply values.yaml]: Added this

**Review by tvoran — APPROVED** [review]
Thanks!

**Pull request merged**